src/daft-core/src/array/ops/shift.rs

```diff
@@ -0,0 +1,127 @@
+use common_error::{DaftError, DaftResult};
+use num_traits::{NumCast, PrimInt};
+
+use crate::{
+    array::DataArray,
+    datatypes::{DaftIntegerType, UInt64Type},
+};
+
+use super::as_arrow::AsArrow;
+
+impl<T> DataArray<T>
+where
+    T: DaftIntegerType,
+    T::Native: PrimInt,
+{
+    pub fn shift_left(&self, bits: &DataArray<UInt64Type>) -> DaftResult<Self> {
+        let result = match (self.len(), bits.len()) {
```

Comment:
You can take some inspiration from this `binary_helper` function here: https://github.com/Eventual-Inc/Daft/blob/main/src/daft-core/src/array/ops/apply.rs#L32 to deal with the broadcasting and validity logic!